inventory/views.py

```diff
@@ -0,0 +1,325 @@
+    @transaction.atomic
+    def form_valid(self, form: BaseModelForm) -> HttpResponse:
+        menuitem = form.cleaned_data.get("menuitem")
+        form.instance.total_cost = menuitem.get_total_cost()
+        form.instance.sale_price = menuitem.price
+        form.instance.user = self.request.user
+        recipe_requirements = RecipeRequirement.objects.filter(menuitem=menuitem)
+        for requirement in recipe_requirements:
+            ingredient = requirement.ingredient
+            if ingredient.quantity_available < requirement.ingredient_quantity:
+                messages.error(self.request, f"Not enough {ingredient.name} to make this menu item.")
+                return super().form_invalid(form)
+            else:
+                ingredient.quantity_available = F("quantity_available") - requirement.ingredient_quantity
+                ingredient.save()
+        return super().form_valid(form)
+
+    # Allow access to MenuItem instance and associated attributes 
+    def get_context_data(self, **kwargs: Any) -> Dict[str, Any]:
+        context = super().get_context_data(**kwargs)
+        form = context.get("form")
+        if form.is_bound and form.is_valid():
+            menuitem = form.cleaned_data.get("menuitem")
+            if menuitem:
+                context["menuitem"] = menuitem
+                context["total_cost"] = menuitem.get_total_cost()
+                context["sale_price"] = menuitem.price
+        return context
+
+class PurchaseUpdate(LoginRequiredMixin, UpdateView):
+    model = Purchase
+    template_name = "inventory/purchase_update.html"
+    fields = ["menuitem"]
+    success_url = reverse_lazy("purchase_log")
+
+class PurchaseDelete(LoginRequiredMixin, DeleteView):
+    model = Purchase
+    template_name = "inventory/purchase_delete.html"
+    success_url = reverse_lazy("purchase_log")
+
+
+
+# REVENUE VIEWS
+class RevenueView(LoginRequiredMixin, TemplateView):
+    template_name = "inventory/revenue.html"
+
+    def get_context_data(self, **kwargs: Any) -> Dict[str, Any]:
+        context = super().get_context_data(**kwargs)
+        # Daily revenue using helper method
+        daily_revenue_dates = Purchase.objects.annotate(date=TruncDate("purchase_time")).values("date").order_by("-date").distinct()
+        daily_revenue = {date["date"]: calculate_daily_revenue(date["date"]) for date in daily_revenue_dates}
+        context["daily_revenue"] = daily_revenue
+        # Monthly revenue using helper method
+        monthly_revenue_months = Purchase.objects.annotate(month=TruncMonth("purchase_time")).values("month").order_by("-month").distinct()
+        monthly_revenue = {month["month"]: calculate_monthly_revenue(month["month"]) for month in monthly_revenue_months}
+        context["monthly_revenue"] = monthly_revenue
+        # Total revenue using helper method 
+        total_revenue = calculate_total_revenue()
+        context["total_revenue"] = total_revenue
+        return context
```

Comment:
Might be a better way to do this but it hasnt immediately jumped out to me (shrug)